tokio/src/io/util/copy.rs

```diff
@@ -36,25 +36,35 @@ impl CopyBuffer {
         W: AsyncWrite + ?Sized,
     {
         loop {
-            // If our buffer is empty, then we need to read some data to
+            // If our buffer has remaining space, then we can read some data to
             // continue.
-            if self.pos == self.cap && !self.read_done {
+            if self.filled < self.buf.len() && !self.read_done {
+                let filled = self.filled;
                 let me = &mut *self;
-                let mut buf = ReadBuf::new(&mut me.buf);
+                let mut buf = ReadBuf::new(&mut me.buf[filled..]);
                 ready!(reader.as_mut().poll_read(cx, &mut buf))?;
```

Comment:
This seems incorrect. If the `poll_read` returns `Pending`, then we don't ever get to try to write anything.